internal/controller/bmc_controller.go

```diff
@@ -193,10 +244,255 @@ func (r *BMCReconciler) discoverServers(ctx context.Context, log logr.Logger, bm
+	running, err := r.checkResetJobStatus(ctx, log, bmcObj)
+	if err != nil {
+		return fmt.Errorf("BMC reset job status: %w", err)
+	}
+	if running {
+		// job is still running, requeue
+		return fmt.Errorf("BMC reset job is still running")
+	}
+	if err := r.updateConditions(ctx, bmcObj, true, bmcResetConditionType, corev1.ConditionTrue, reason, message); err != nil {
+		return fmt.Errorf("failed to set BMC resetting condition: %w", err)
+	}
+	bmcClient, err := bmcutils.GetBMCClientFromBMC(ctx, r.Client, bmcObj, r.Insecure, r.BMCPollingOptions)
+	if err == nil {
+		if err := bmcClient.ResetManager(ctx, bmcObj.Spec.BMCUUID, redfish.GracefulRestartResetType); err == nil {
+			log.Info("Successfully reset BMC via Redfish", "BMC", bmcObj.Name)
+			return nil
+		}
+		log.Error(err, "failed to reset BMC via Redfish, falling back to rest via ssh", "BMC", bmcObj.Name)
+	}
+	if httpErr, ok := err.(*common.Error); ok {
+		// only handle 5xx errors
+		if httpErr.HTTPReturnedStatusCode < 500 || httpErr.HTTPReturnedStatusCode >= 600 {
+			return fmt.Errorf("cannot reset bmc: %w", err)
+		}
+	} else {
+		return fmt.Errorf("cannot reset bmc, unknown error: %w", err)
+	}
+	// If we reach here, Redfish reset failed or was not possible, fall back to reset via ssh
+	newJob := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("bmc-reset-%s", bmcObj.Name),
+			Namespace: r.ManagerNamespace,
+			Labels:    bmcObj.Labels,
+		},
+		Spec: batchv1.JobSpec{
+			BackoffLimit: func(i int32) *int32 { return &i }(3),
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					RestartPolicy: corev1.RestartPolicyNever,
+					Containers: []corev1.Container{
+						{
+							Name:  "bmc-reset",
+							Image: r.BMCToolsImage,
+							Command: []string{
+								"bmc",
+								"reset",
+								bmcObj.Name,
+								"--model", bmcObj.Status.Model,
+								"--bmc_address", bmcObj.Spec.Endpoint.IP.String(),
+								"--bmc_manufacturer", bmcObj.Status.Manufacturer,
+								"--timeout", "5m",
+							},
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Env: []corev1.EnvVar{
+								{
+									Name: metalv1alpha1.BMCSecretPasswordKeyName,
+									ValueFrom: &corev1.EnvVarSource{
+										SecretKeyRef: &corev1.SecretKeySelector{
+											Key: metalv1alpha1.BMCSecretPasswordKeyName,
+											LocalObjectReference: corev1.LocalObjectReference{
+												Name: bmcObj.Spec.BMCSecretRef.Name,
+											},
+										},
+									},
+								},
+								{
+									Name: metalv1alpha1.BMCSecretUsernameKeyName,
+									ValueFrom: &corev1.EnvVarSource{
+										SecretKeyRef: &corev1.SecretKeySelector{
+											Key: metalv1alpha1.BMCSecretPasswordKeyName,
+											LocalObjectReference: corev1.LocalObjectReference{
+												Name: bmcObj.Spec.BMCSecretRef.Name,
+											},
+										},
+									},
+								},
```

Comment:
The referenced keys need to reside in `corev1.Secret` resources to be available as env vars. The referenced `BMCSecrets` are `metalv1alphav1.BMCSecret` resource, which cannot be used that way.